nucypher/characters/utils.py

```diff
@@ -0,0 +1,113 @@
+from random import shuffle
+
+import maya
+from umbral.keys import UmbralPublicKey
+
+from nucypher.acumen.perception import FleetSensor
+from nucypher.crypto.signing import InvalidSignature
+from nucypher.network.exceptions import NodeSeemsToBeDown
+from nucypher.network.nodes import Learner
+
+
+def get_treasure_map(learner: Learner, map_identifier: str, bob_encrypting_key: UmbralPublicKey, timeout=3):
+    """
+    Iterate through the nodes we know, asking for the TreasureMap.
+    Return the first one who has it.
+    """
+    if learner.federated_only:
+        from nucypher.policy.collections import TreasureMap as _MapClass
+    else:
+        from nucypher.policy.collections import SignedTreasureMap as _MapClass
+
+    start = maya.now()
+
+    # Spend no more than half the timeout finding the nodes.  8 nodes is arbitrary.  Come at me.
+    learner.block_until_number_of_known_nodes_is(8, timeout=timeout / 2, learn_on_this_thread=True)
+    while True:
+        nodes_with_map = matching_nodes_among(nodes=learner.known_nodes,
+                                              bob_encrypting_key=bob_encrypting_key)
+        # TODO nodes_with_map can be large - what if treasure map not present in any of them?
+        shuffle(nodes_with_map)
+
+        for node in nodes_with_map:
+            try:
+                response = learner.network_middleware.get_treasure_map_from_node(node, map_identifier)
+            except (*NodeSeemsToBeDown, learner.NotEnoughNodes):
+                continue
+            except learner.network_middleware.NotFound:
+                learner.log.info(f"Node {node} claimed not to have TreasureMap {map_identifier}")
+                continue
+            except node.NotStaking:
+                # TODO this wasn't here before - check with myles
+                learner.log.info(f"Node {node} not staking")
+                continue
+
+            if response.status_code == 200 and response.content:
+                try:
+                    treasure_map = _MapClass.from_bytes(response.content)
+                    return treasure_map
+                except InvalidSignature:
+                    # TODO: What if a node gives a bunk TreasureMap?  NRN
+                    raise
+            else:
+                continue  # TODO: Actually, handle error case here.  NRN
+        else:
+            learner.learn_from_teacher_node()
+
+        if (start - maya.now()).seconds > timeout:
+            raise _MapClass.NowhereToBeFound(f"Asked {len(learner.known_nodes)} nodes, "
+                                             f"but none had map {map_identifier}")
+
+
+def matching_nodes_among(nodes: FleetSensor,
+                         bob_encrypting_key: UmbralPublicKey,
+                         no_less_than=7):  # Somewhat arbitrary floor here.
+    # Look for nodes whose checksum address has the second character of Bob's encrypting key in the first
+    # few characters.
+    # Think of it as a cheap knockoff hamming distance.
```

Comment:
These are nodes picked for the treasure map distribution which is different from nodes picked for policies. The "hamming distance" algorithm was derived by @jMyles , and makes finding treasure maps on the network faster by being more deterministic; Alice has to distribute the treasure map on the network, and then Bob has to find it on the network - if they both use the same algorithm for deciding which nodes to use/check with then the process is more efficient. 

This was previous functionality that I simply moved to be in a common location so that both Characters and Porter could use the same code.

In the TMapConKFrags world, distributing the treasure map would not be required, and Bob would just store it himeself.